Feeder audit 2026-05-29: nve-hydro

## Feeder audit summary for `nve-hydro`

This sweep covers the objective repo-state parts of the feeder-release-checklist: required file presence, catalog + ghpages source entries, source-local pytest with generated producer src paths injected into PYTHONPATH, Docker E2E matrix coverage, ARM static audit, Fabric static audit, and ARM generator dry-run no-diff. It does not rerun live Azure/Fabric deploy bullets, specialist-agent reviews, or upstream API re-audits.

## Findings

- **ERROR** `missing-docker-build-matrix-entry` — 1 finding(s)
  - tests/docker_e2e/matrix.json has no build entry for kafka (Dockerfile). (tests/docker_e2e/matrix.json)
- **ERROR** `missing-docker-flow-matrix-entry` — 1 finding(s)
  - tests/docker_e2e/matrix.json has no flow entry for kafka (Dockerfile). (tests/docker_e2e/matrix.json)
- **ERROR** `missing-file` — 2 finding(s)
  - Required file missing: feeders/nve-hydro/xreg/nve-hydro.xreg.json (feeders/nve-hydro/xreg/nve-hydro.xreg.json)
  - Required file missing: feeders/nve-hydro/kql/nve-hydro.kql (feeders/nve-hydro/kql/nve-hydro.kql)
- **ERROR** `pytest-failed` — 1 finding(s)
  - Source-local pytest failed. (feeders/nve-hydro/tests)
- **WARNING** `arm-stub-description` — 3 finding(s)
  - azure-template-with-eventhub.json: parameter 'eventHubSku' has stub or missing metadata.description (feeders/nve-hydro/azure-template-with-eventhub.json)
  - azure-template-with-servicebus.json: parameter 'queueName' has stub or missing metadata.description (feeders/nve-hydro/azure-template-with-servicebus.json)
  - azure-template-amqp.json: parameter 'queueName' has stub or missing metadata.description (feeders/nve-hydro/infra/azure-template-amqp.json)

## Check evidence

### `arm_audit` — rc=0, 1.31s

```text
ARM template static audit
  RepoRoot:        C:\git\real-time-sources
  Feeders scanned: 1
  Templates:       6
  Blockers:        0
  Warnings:        3

[nve-hydro] blockers=0 warnings=3
  azure-template-with-eventhub.json
    [WARNING] stub-description: parameter 'eventHubSku' has stub or missing metadata.description
  azure-template-with-servicebus.json
    [WARNING] stub-description: parameter 'queueName' has stub or missing metadata.description
  azure-template-amqp.json
    [WARNING] stub-description: parameter 'queueName' has stub or missing metadata.description
PASS (with 3 warning(s)).
Report written to C:\Users\clemensv\.copilot\session-state\cb5ea344-87f5-406e-a13a-a66811763074\files\fleet-release-audit\arm\nve-hydro.json
```

### `arm_generator` — rc=0, 2.75s

```text
Description sources: total=177, container=31, readme=0, argparse=28, fallback=118
Generated 0 ARM template(s).
```

### `fabric_audit` — rc=0, 1.03s

```text
[32;1mSource   [0m[32;1m OptIn[0m[32;1m Notebook[0m[32;1m  Hook[0m[32;1m PASS[0m[32;1m WARN[0m[32;1m BLOCK[0m
[32;1m------   [0m [32;1m-----[0m [32;1m--------[0m [32;1m ----[0m [32;1m----[0m [32;1m----[0m [32;1m-----[0m
nve-hydro  True     True False   10    0     0

Sources: 1 total, 0 blockers, 0 warnings
JSON report: C:\Users\clemensv\.copilot\session-state\cb5ea344-87f5-406e-a13a-a66811763074\files\fleet-release-audit\fabric\nve-hydro.json
```

### `pytest` — rc=1, 4.94s

```text
_CatchAllUndefinedParameters._separate_defined_undefined_kvs(
                obj, kwargs)
        num_params_takeable = len(
            init_signature.parameters) - 1  # don't count self
        num_args_takeable = num_params_takeable - len(known_kwargs)
    
        args = args[:num_args_takeable]
        bound_parameters = init_signature.bind_partial(self, *args,
                                                       **known_kwargs)
        bound_parameters.apply_defaults()
    
        arguments = bound_parameters.arguments
        arguments.pop("self", None)
        final_parameters = \
            _IgnoreUndefinedParameters.handle_from_dict(obj, arguments)
>       original_init(self, **final_parameters)
E       TypeError: WaterLevelObservation.__init__() missing 1 required keyword-only argument: 'river_name'

C:\Users\clemensv\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\dataclasses_json\undefined.py:117: TypeError
============================== warnings summary ===============================
..\..\..\..\Users\clemensv\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\avro\schema.py:1229
  C:\Users\clemensv\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\avro\schema.py:1229: IgnoredLogicalType: Logical type timestamp-millis requires literal type long, not string.
    warnings.warn(

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html

---------- coverage: platform win32, python 3.11.9-final-0 -----------
Name                     Stmts   Miss  Cover   Missing
------------------------------------------------------
nve_hydro\__init__.py        2      0   100%
nve_hydro\__main__.py        2      2     0%   1-2
nve_hydro\nve_hydro.py     219    121    45%   92-93, 106-107, 135-168, 174-228, 232-301, 305
------------------------------------------------------
TOTAL                      223    123    45%

=========================== short test summary info ===========================
FAILED tests/test_nve_hydro_unit.py::TestDataClasses::test_observation_roundtrip
!!!!!!!!!!!!!!!!!!!!!!!!!! stopping after 1 failures !!!!!!!!!!!!!!!!!!!!!!!!!!
1 failed, 27 passed, 1 warning in 1.16s
```

## Next step

Fix the findings above, then rerun the sequential feeder-release-checklist audit for this feeder and close this issue with the new evidence.

--- a/feeders/nve-hydro/tests/test_nve_hydro_unit.py
+++ b/feeders/nve-hydro/tests/test_nve_hydro_unit.py
@@ -290,6 +290,7 @@ class TestDataClasses:
     def test_observation_roundtrip(self):
         obs = WaterLevelObservation(
             station_id="2.11.0",
+            river_name="Vosso",
             water_level=1.45,
             water_level_unit="m",
             water_level_timestamp="2023-11-14T13:00:00Z",
@@ -299,6 +300,7 @@ class TestDataClasses:
         )
         assert obs.station_id == "2.11.0"
         data = json.loads(obs.to_json())
+        assert data["river_name"] == "Vosso"
         assert data["water_level"] == 1.45
         assert data["discharge"] == 45.6
 
@@ -318,21 +320,23 @@ class TestSendStations:
         mock_gs.return_value = [SAMPLE_STATION_1, SAMPLE_STATION_2]
         api = NVEHydroAPI("key")
         prod = self._mock_producer()
-        station_params = send_stations(api, prod)
+        station_params, station_river = send_stations(api, prod)
         assert prod.send_no_nve_hydrology_station.call_count == 2
         prod.producer.flush.assert_called_once()
         assert PARAM_STAGE in station_params["2.11.0"]
         assert PARAM_DISCHARGE in station_params["2.11.0"]
         assert PARAM_STAGE in station_params["12.193.0"]
         assert PARAM_DISCHARGE not in station_params["12.193.0"]
+        assert station_river == {"2.11.0": "Vosso", "12.193.0": "Nidelva"}
 
     @patch.object(NVEHydroAPI, "get_stations")
     def test_skips_station_without_id(self, mock_gs):
         mock_gs.return_value = [{"stationName": "No ID station", "seriesList": []}]
         api = NVEHydroAPI("key")
         prod = self._mock_producer()
-        result = send_stations(api, prod)
-        assert result == {}
+        station_params, station_river = send_stations(api, prod)
+        assert station_params == {}
+        assert station_river == {}
         prod.send_no_nve_hydrology_station.assert_not_called()
 
     @patch.object(NVEHydroAPI, "get_stations")
@@ -340,8 +344,9 @@ class TestSendStations:
         mock_gs.return_value = []
         api = NVEHydroAPI("key")
         prod = self._mock_producer()
-        result = send_stations(api, prod)
-        assert result == {}
+        station_params, station_river = send_stations(api, prod)
+        assert station_params == {}
+        assert station_river == {}
 
 
 # ---------------------------------------------------------------------------
@@ -367,8 +372,9 @@ class TestFeedObservations:
         api = NVEHydroAPI("key")
         prod = self._mock_producer()
         station_params = {"2.11.0": [PARAM_STAGE, PARAM_DISCHARGE]}
+        station_river = {"2.11.0": "Vosso"}
         previous = {}
-        count = feed_observations(api, prod, station_params, previous)
+        count = feed_observations(api, prod, station_params, station_river, previous)
         assert count == 1
         prod.send_no_nve_hydrology_water_level_observation.assert_called_once()
         prod.producer.flush.assert_called()
@@ -381,11 +387,12 @@ class TestFeedObservations:
         api = NVEHydroAPI("key")
         prod = self._mock_producer()
         station_params = {"2.11.0": [PARAM_STAGE]}
+        station_river = {"2.11.0": "Vosso"}
         previous = {}
-        feed_observations(api, prod, station_params, previous)
+        feed_observations(api, prod, station_params, station_river, previous)
         prod.reset_mock()
         prod.producer = MagicMock()
-        count = feed_observations(api, prod, station_params, previous)
+        count = feed_observations(api, prod, station_params, station_river, previous)
         assert count == 0
 
     @patch.object(NVEHydroAPI, "get_observations")
@@ -393,7 +400,7 @@ class TestFeedObservations:
         mock_obs.return_value = []
         api = NVEHydroAPI("key")
         prod = self._mock_producer()
-        count = feed_observations(api, prod, {"2.11.0": [PARAM_STAGE]}, {})
+        count = feed_observations(api, prod, {"2.11.0": [PARAM_STAGE]}, {"2.11.0": "Vosso"}, {})
         assert count == 0
 
 
@@ -419,6 +426,7 @@ class TestProducerClient:
         prod = NONVEHydrologyEventProducer(mock_kafka, "nve-topic")
         obs = WaterLevelObservation(
             station_id="2.11.0",
+            river_name="Vosso",
             water_level=1.45, water_level_unit="m",
             water_level_timestamp="2023-11-14T13:00:00Z",
             discharge=None, discharge_unit="m3/s", discharge_timestamp="",

--- a/feeders/nve-hydro/tests/test_once_mode.py
+++ b/feeders/nve-hydro/tests/test_once_mode.py
@@ -12,7 +12,7 @@ def fake_api_and_producer():
     with patch('nve_hydro.nve_hydro.NVEHydroAPI') as api_cls, \
          patch('nve_hydro.nve_hydro.Producer') as producer_cls, \
          patch('nve_hydro.nve_hydro.NONVEHydrologyEventProducer') as nve_prod_cls, \
-         patch('nve_hydro.nve_hydro.send_stations', return_value={}) as send_st, \
+         patch('nve_hydro.nve_hydro.send_stations', return_value=({}, {})) as send_st, \
          patch('nve_hydro.nve_hydro.feed_observations', return_value=0) as feed_obs, \
          patch('nve_hydro.nve_hydro._load_state', return_value={}), \
          patch('nve_hydro.nve_hydro._save_state'), \

